File: OneClickMiner.Tests/FileLoggerTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FakeItEasy;
using NUnit.Framework;
using VertcoinOneClickMiner.Core;

namespace OneClickMiner.Tests
{
    [TestFixture]
    class FileLoggerTests
    {
        private string logFilePath;
        private ISystemWrapper systemIo;

        private FileLogger CreateSut()
        {
            return new FileLogger(logFilePath, systemIo);
        }

        [SetUp]
        public void SetUp()
        {
            logFilePath = "theLogFile";
            systemIo = A.Fake<ISystemWrapper>();
        }

        [Test]
        public void Trace_WhenCalled_ShouldAppendToLogFileWithExpectedFormat()
        {
            string resultLogFile = null;
            IEnumerable<string> resultLogLines = null;

            var timestamp = DateTime.Now;
            A.CallTo(() => systemIo.UtcNow).Returns(timestamp);
            A.CallTo(() => systemIo.AppendAllLines(A<string>.Ignored, A<IEnumerable<string>>.Ignored))
                .Invokes((string file, IEnumerable<string> lines) =>
                {
                    resultLogFile = file;
                    resultLogLines = lines;
                });

            var sut = CreateSut();

            string message = "the log message";
            string caller = "the method which called the logger";

            sut.Trace(message, caller);
            var expectedLogString = $"- {timestamp:O}, {caller}: {message}";
            
            Assert.That(resultLogFile, Is.EqualTo(logFilePath));
            Assert.That(resultLogLines.First(), Is.EqualTo(expectedLogString));
        }

        [Test]
        public void LogError_WhenCalled_ShouldAppendToLogFileWithExceptionDetails()
        {
            string resultLogFile = null;
            IEnumerable<string> resultLogLines = null;

            var timestamp = DateTime.Now;
            A.CallTo(() => systemIo.UtcNow).Returns(timestamp);
            A.CallTo(() => systemIo.AppendAllLines(A<string>.Ignored, A<IEnumerable<string>>.Ignored))
                .Invokes((string file, IEnumerable<string> lines) =>
                {
                    resultLogFile = file;
                    resultLogLines = lines;
                });

            var sut = CreateSut();

            var exception = new Exception("something went bang");

            string message = exception.ToString();
            string caller = "the caller method";

            sut.LogError(exception, caller);
            var expectedLogString = $"- {timestamp:O}, {caller}: {message}";
            
            Assert.That(resultLogFile, Is.EqualTo(logFilePath));
            Assert.That(resultLogLines.First(), Is.EqualTo(expectedLogString));
        }
    }
}

File: OneClickMiner.Tests/FileLoggerTests.vb
Imports FakeItEasy
Imports NUnit.Framework
Imports VertcoinOneClickMiner.Core


<TestFixture> _
Class FileLoggerTests
    Private logFilePath As String
    Private systemIo As ISystemWrapper

    Private Function CreateSut() As FileLogger
        Return New FileLogger(logFilePath, systemIo)
    End Function

    <SetUp> _
    Public Sub SetUp()
        logFilePath = "theLogFile"
        systemIo = A.Fake(Of ISystemWrapper)()
    End Sub

    <Test>
    Public Sub Constructor_WhenCalledWithNullFilePath_ShouldThrowArgumentNullException()

        logFilePath = Nothing

        Assert.Throws(of ArgumentNullException)(Sub() CreateSut())

    End Sub

    <Test> _
    <TestCase("")> _
    <TestCase(vbCrLf)> _
    <TestCase(" ")> _
    Public Sub Constructor_WhenCalledWithInvalidFilePath_ShouldThrowArgumentException(path As string)

        logFilePath = path

        Assert.Throws(of ArgumentException)(Sub() CreateSut())

    End Sub

    <Test> _
    Public Sub Trace_WhenCalled_ShouldAppendToLogFileWithExpectedFormat()
        Dim resultLogFile As String = Nothing
        Dim resultLogLines As IEnumerable(Of String) = Nothing

        Dim timestamp = DateTime.Now
        A.CallTo(Function() systemIo.UtcNow).Returns(timestamp)
        A.CallTo(Sub() systemIo.AppendAllLines(A(Of string).Ignored, A(of IEnumerable(Of string)).Ignored)) _
            .Invokes(Sub(file As String, lines As IEnumerable(Of String))
                resultLogFile = file
                resultLogLines = lines
                        End Sub)

        Dim sut = CreateSut()

        Dim message As String = "the log message"
        Dim caller As String = "the method which called the logger"

        sut.Trace(message, caller)
        Dim expectedLogString = $"- {timestamp:O}, {caller}: {message}"

        Assert.That(resultLogFile, [Is].EqualTo(logFilePath))
        Assert.That(resultLogLines.First(), [Is].EqualTo(expectedLogString))
    End Sub

    <Test> _
    Public Sub LogError_WhenCalled_ShouldAppendToLogFileWithExceptionDetails()
        Dim resultLogFile As String = Nothing
        Dim resultLogLines As IEnumerable(Of String) = Nothing

        Dim timestamp = DateTime.Now
        A.CallTo(Function() systemIo.UtcNow).Returns(timestamp)
        A.CallTo(Sub() systemIo.AppendAllLines(A(Of string).Ignored, A(of IEnumerable(Of string)).Ignored)) _
            .Invokes(Sub(file As String, lines As IEnumerable(Of String))
                resultLogFile = file
                resultLogLines = lines
                        End Sub)

        Dim sut = CreateSut()

        Dim exception = New Exception("something went bang")

        Dim message As String = exception.ToString()
        Dim caller As String = "the caller method"

        sut.LogError(exception, caller)
        Dim expectedLogString = $"- {timestamp:O}, {caller}: {message}"

        Assert.That(resultLogFile, [Is].EqualTo(logFilePath))
        Assert.That(resultLogLines.First(), [Is].EqualTo(expectedLogString))
    End Sub
End Class